### PR TITLE
Debug tpose

### PR DESCRIPTION
**Pull request by Gianlucamilani**
In this issue we are going to fix the tpose bug observed in https://github.com/ami-iit/component_ifeel/issues/1609

### CODE REVIEW

File: src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h [hunk 1]
```diff
@@ -42,37 +42,32 @@ class HumanIK
...
-        const std::string& taskName,
-        const std::shared_ptr<BipedalLocomotion::ParametersHandler::IParametersHandler> taskHandler);
+    bool initializeJointRegularizationTask(const std::string& taskName,
+                                           const std::shared_ptr<BipedalLocomotion::ParametersHandler::IParametersHandler> taskHandler);
 
-    bool initializeJointConstraintsTask(
-        const std::string& taskName,
-        const std::shared_ptr<BipedalLocomotion::ParametersHandler::IParametersHandler> taskHandler);
+    bool initializeJointConstraintsTask(const std::string& taskName,
+                                        const std::shared_ptr<BipedalLocomotion::ParametersHandler::IParametersHandler> taskHandler);
```
Comment (GiulioRomualdi) [inline src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h:70]: I would add the documentation for these

File: src/IK/src/InverseKinematics.cpp [hunk 2]
```diff
@@ -33,16 +32,10 @@ bool HumanIK::initialize(
     m_jointVelocities.resize(m_kinDyn->getNrOfDegreesOfFreedom());
 
     // Retrieve the state of the system
-    kinDyn->getRobotState(m_basePose,
-                          m_jointPositions,
-                          m_baseVelocity,
-                          m_jointVelocities,
-                          m_gravity);
+    kinDyn->getRobotState(m_basePose, m_jointPositions, m_baseVelocity, m_jointVelocities, m_gravity);
```
Comment (GiulioRomualdi) [inline src/IK/src/InverseKinematics.cpp:35]: You can check if the the function return true

File: src/IK/src/InverseKinematics.cpp [hunk 3]
```diff
@@ -272,28 +243,25 @@ bool HumanIK::updateJointConstraintsTask()
...
 bool HumanIK::TPoseCalibrationNode(const int node, const manif::SO3d& I_R_IMU)
 {
+    tPose = true;
     // check if the node number is valid
-    if ((m_OrientationTasks.find(node) == m_OrientationTasks.end())
-        && (m_GravityTasks.find(node) == m_GravityTasks.end()))
+    if ((m_OrientationTasks.find(node) == m_OrientationTasks.end()) && (m_GravityTasks.find(node) == m_GravityTasks.end()))
     {
-        BiomechanicalAnalysis::log()->error("[HumanIK::TPoseCalibrationNode] Invalid node number.");
+        BiomechanicalAnalysis::log()->error("[HumanIK::setNodeSetPoint] Invalid node number.");
```
Comment (GiulioRomualdi) [inline src/IK/src/InverseKinematics.cpp:250]:
```suggestion
        BiomechanicalAnalysis::log()->error("[HumanIK::TPoseCalibrationNode] Invalid node number.");
```

File: src/IK/src/InverseKinematics.cpp [hunk 4]
```diff
@@ -852,6 +794,8 @@ bool HumanIK::initializeJointConstraintsTask(
         {
             lowerLimits[i] = m_kinDyn->model().getJoint(i)->getMinPosLimit(i);
             upperLimits[i] = m_kinDyn->model().getJoint(i)->getMaxPosLimit(i);
+            std::cout << "joint " << m_kinDyn->model().getJointName(i) << " upper lim " << upperLimits[i] << " lower limit "
+                      << lowerLimits[i] << " joint pos: " << m_jointPositions[i] << std::endl;
```
Comment (GiulioRomualdi) [inline src/IK/src/InverseKinematics.cpp:798]: I think this can be removed

File: src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h [hunk 5]
```diff
@@ -174,6 +167,7 @@ class HumanIK
     object */
 
     int m_nrDoFs; /** Number of Joint Degrees of Freedom */
+    bool tPose{false}; /** Flag for resetting the integrator state */
```
Comment (davidegorbani) [inline src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h:170]:
```suggestion
    bool m_tPose{false}; /** Flag for resetting the integrator state */
```

I would rename this variable to be consistent with the rest of the code by adding `m_` in front.

File: src/IK/src/InverseKinematics.cpp [hunk 6]
```diff
@@ -178,27 +157,22 @@ int HumanIK::getDoFsNumber() const
...
-bool HumanIK::updateOrientationTask(const int node,
-                                    const manif::SO3d& I_R_IMU,
-                                    const manif::SO3Tangentd& I_omega_IMU)
+bool HumanIK::updateOrientationTask(const int node, const manif::SO3d& I_R_IMU, const manif::SO3Tangentd& I_omega_IMU)
 {
     // Check if the node number is valid
     if (m_OrientationTasks.find(node) == m_OrientationTasks.end())
     {
-        BiomechanicalAnalysis::log()->error("[HumanIK::setNodeSetPoint] Invalid node number.");
+        BiomechanicalAnalysis::log()->error("[HumanIK::TPoseCalibrationNode] Invalid node number.");
```
Comment (davidegorbani) [inline src/IK/src/InverseKinematics.cpp:165]:
```suggestion
        BiomechanicalAnalysis::log()->error("[HumanIK::updateOrientationTask] Invalid node number.");
```

File: src/examples/IK/exampleIK.cpp [hunk 7]
```diff
@@ -273,177 +262,137 @@ int main()
...
+    initialJointPositions.resize(kinDyn->getNrOfDegreesOfFreedom()); // Resize the initial joint pos vec based on the n. of deg of freedom
+    initialJointVelocities.resize(kinDyn->getNrOfDegreesOfFreedom()); // Resize the initial joint vel vec based on the n. of deg of freedom
+
+    //*************************************************************************************************************************************
+
+    //*************************************************************************************************************************************
+    // Load kinematic data from Matlab file human_data
+    //*************************************************************************************************************************************
+
+    matioCpp::File file2("/path/to/human_data1.mat"); // Create another file object to read data from
```
Comment (davidegorbani) [inline src/examples/IK/exampleIK.cpp:307]: maybe here we can read the path to the matlab file from the [ini file](https://github.com/ami-iit/biomechanical-analysis-framework/blob/debug-tpose/src/examples/IK/exampleIK.ini).

File: src/examples/IK/exampleIK.cpp [hunk 8]
```diff
@@ -273,177 +262,137 @@ int main()
...
-                                                                      // structure and convert it to
-                                                                      // a multidimensional array of
-                                                                      // type double
+        = base_orientation("data").asMultiDimensionalArray<double>(); // Access multidimensional array "data" within "base_orientation"
+                                                                      // structure and convert it to a multidimensional array of type double
+
+    //*************************************************************************************************************************************
+
+    // Get the dimension of the data (total number of timestamps) from Matlab file
+    matioCpp::File file("/path/to/matlab1.mat"); // Create a file object to read data from MATLAB file
```
Comment (davidegorbani) [inline src/examples/IK/exampleIK.cpp:345]: And here the same.

**Review by davidegorbani — COMMENTED** [review]
I left a couple of suggestions.

File: src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h (hunk 1 above)
Comment (Gianlucamilani) [reply src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h]: Done with commit https://github.com/ami-iit/biomechanical-analysis-framework/pull/31/commits/2a4606a02a2dcf6ed2fa2564db8d7b83456d2c40

File: src/IK/src/InverseKinematics.cpp (hunk 4 above)
Comment (Gianlucamilani) [reply src/IK/src/InverseKinematics.cpp]: Done with commit https://github.com/ami-iit/biomechanical-analysis-framework/pull/31/commits/2a4606a02a2dcf6ed2fa2564db8d7b83456d2c40

File: src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h (hunk 5 above)
Comment (Gianlucamilani) [reply src/IK/include/BiomechanicalAnalysis/IK/InverseKinematics.h]: Done with commit https://github.com/ami-iit/biomechanical-analysis-framework/pull/31/commits/575da1ffecd6145855cad2d5019c3e90325b45b3

**Review by davidegorbani — APPROVED** [review]

File: src/IK/src/InverseKinematics.cpp (hunk 6 above)
Comment (Gianlucamilani) [reply src/IK/src/InverseKinematics.cpp]: Done with commit https://github.com/ami-iit/biomechanical-analysis-framework/pull/31/commits/575da1ffecd6145855cad2d5019c3e90325b45b3

File: src/examples/IK/exampleIK.cpp (hunk 7 above)
Comment (Gianlucamilani) [reply src/examples/IK/exampleIK.cpp]: Done with commit https://github.com/ami-iit/biomechanical-analysis-framework/pull/31/commits/575da1ffecd6145855cad2d5019c3e90325b45b3

File: src/examples/IK/exampleIK.cpp (hunk 8 above)
Comment (Gianlucamilani) [reply src/examples/IK/exampleIK.cpp]: Done with commit https://github.com/ami-iit/biomechanical-analysis-framework/pull/31/commits/575da1ffecd6145855cad2d5019c3e90325b45b3

**Comment by Gianlucamilani** [comment]
All comments have been addressed with @davidegorbani , thanks

cc @GiulioRomualdi

**Review by traversaro — APPROVED** [review]

**Comment by GiulioRomualdi** [comment]
Thank you @Gianlucamilani !

**Comment by davidegorbani** [comment]
Merging! Thank you @Gianlucamilani @GiulioRomualdi @traversaro!

**Pull request merged**